cli-plugins/manager/manager.go

```diff
@@ -0,0 +1,190 @@
+package manager
+
+import (
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"runtime"
+	"strings"
+	"sync"
+	"syscall"
+
+	"github.com/docker/cli/cli/command"
+	"github.com/spf13/cobra"
+)
+
+// ErrPluginNotFound is the error returned when a plugin could not be found.
+type ErrPluginNotFound string
+
+func (e ErrPluginNotFound) Error() string {
+	return "Error: No such CLI plugin: " + string(e)
+}
+
+var (
+	pluginDirs     []string
+	pluginDirsOnce sync.Once
+)
+
+func getPluginDirs(dockerCli command.Cli) []string {
+	pluginDirsOnce.Do(func() {
+		// Mostly for test.
+		if ds := os.Getenv("DOCKER_CLI_PLUGIN_EXTRA_DIRS"); ds != "" {
```

Comment:
I'm not very comfortable with this code added only for tests. I'm concerned with security issues it might introduce, as it is loaded first, and might shadow regular plugins with corrupted ones? And as it is an undocumented environment variable and a non standard way to configure cli plugin directories, basic user will never notice it if it is set by someone else...

I think you need this env var only for e2e tests, but maybe you can transform you e2e test as a unit test, in this very package, from where then you can "patch" the `pluginsDirs` variable (even if it's not a good way too but somehow better I guess)?

Reply:
Despite the comment I'm actually considering whether this should perhaps be a user visible/usable/documented setting.

I don't think the e2e tests which need this can become unit tests, they need to actually run the CLI and see that it finds plugins etc.

Reply:
I thought I said this already, but I can't see where: I'm going to try and nuke the envvar by using `docker -C somecustomplace` in the e2e tests.